CORE-14879: Migrate "Roles" Admin API v1 Security Endpoints to v2 ConnectRPC API

nguyen-andrew opened the pull request:
This PR introduces comprehensive support and testing for Role Management in the Admin API v2.

The main changes include:
- Implementation of the Role Management endpoints for Admin API v2, enabling administrators to get, list, create, and delete roles, as well as manage their members. Write operations are proxied to the controller leader, while read operations can be served locally, consistent with v1 API behavior.
- Generation of Ducktape protobuf clients to test the new Admin API v2 Roles security endpoints.
- Introduction of a new test file for Admin API v2 role management, replicating and extending tests from the original rbac_tests.py to ensure functional coverage.

https://redpandadata.atlassian.net/browse/CORE-14879

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.3.x
- [ ] v25.2.x
- [ ] v25.1.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.
-->
### Features
* Added role management in Admin API v2, including create, list, get, and delete roles, and manage role members.
<!--
### Improvements

* Short description of how this PR improves existing behavior.

-->

nguyen-andrew committed:
admin: Add Role management stubs in Admin API v2

This commit adds the protobuf messages and RPCs for role management for
the Admin API V2, and it extends the security service implementation to
include unimplemented stubs for the new role management methods:
- CreateRole
- GetRole
- ListRoles
- AddRoleMembers
- RemoveRoleMembers
- DeleteRole
- ListCurrentUserRoles

nguyen-andrew committed:
dt/pb: Generate Ducktape protobuf clients

Generate Ducktape protobuf clients to test the new Admin API v2 Roles
security endpoints.

nguyen-andrew committed:
dt: Adding Admin V2 role management API tests

Added a new file to test the Admin API v2 for role management. This file
will be based on the existing rbac_tests.py and will aim to provide the
same level of functional testing and coverage by using the original
rbac_tests.py as a reference. This initial commit will replicate most
of the tests from the RBACTest class for the v2 API (some TODOs remain).

Additionally, updated tools/type-checking/type-check-strictness.json to
include the new rbac_test_v2.py file for type checking.

nguyen-andrew committed:
admin: Implement Role Management Admin API v2

Implemented the Role Management endpoints for the Admin API v2, allowing
administrators to get, list, create, and delete roles as well as manage
their members. All write operations are proxied to the controller leader
while read operations can be served locally. This is consistent with the
existing behavior of the v1 Admin API.

nguyen-andrew committed:
dt: Extending rbac_test_v2 to include RBACEndToEndTest class

Replicating the RBACEndToEndTest class from rbac_test.py to ensure
comprehensive testing of RBAC functionalities in the Admin API v2.

nguyen-andrew committed:
dt: Extending rbac_test_v2 to include RolePersistenceTest class

Replicating the RolePersistenceTest class from rbac_test.py to ensure
comprehensive testing of RBAC functionalities in the Admin API v2.

nguyen-andrew commented on src/v/redpanda/admin/services/security.cc:
```diff
@@ -12,74 +12,458 @@
+          ssx::sformat("Invalid role name {{{}}}", role_name()));
+    }
+
+    // Question: Should we also validate role member names?
```
Should we validate member name as well? V1 doesn't, but wanted to get your thoughts @michael-redpanda

michael-redpanda replied:
Good question - I think the control character check is a good one, and we may be able to do the `validate_scram_username` check against _users_ for roles, but that may not place nicely with groups.  e.g. SASL usernames cna't have "," or "=" but groups may contain those characters.  I think control character checking at a minimum is a good idea though.

nguyen-andrew replied:
Sounds good, will add control character checking for now

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--67465e2c-15fc-4e45-9bdb-09a9ff848dc3-->
<!--TEST RESULTS-->
<!--f93ee5d0-9f69-436c-ad11-afe2be13d5e9-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/78219">build#78219</a></summary>


| test_class               | test_method              | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                    |
|:-------------------------|:-------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| NodesDecommissioningTest | test_decommission_status | null             | integration | https://buildkite.com/redpanda/redpanda/builds/78219#019b37dc-7c56-496b-a8cb-ea807285313c | FLAKY         | 36/41    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0417, p0=0.0841, reject_threshold=0.0100. adj_baseline=0.1199, p1=0.4680, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommission_status |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/78409">build#78409</a></summary>


| test_class               | test_method                                   | test_arguments                                                            | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                                         |
|:-------------------------|:----------------------------------------------|:--------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| MountUnmountIcebergTest  | test_simple_remount                           | {"cloud_storage_type": 1}                                                 | integration | https://buildkite.com/redpanda/redpanda/builds/78409#019b53d4-0da2-4d4b-8042-fed89174e779 | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.1800, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.4486, p1=0.0026, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount                            |
| NodesDecommissioningTest | test_decommissioning_cancel_ongoing_movements | {"cloud_topic": true}                                                     | integration | https://buildkite.com/redpanda/redpanda/builds/78409#019b53d4-0da4-4430-832a-f3c6e5305c41 | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0000, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.1000, p1=0.3487, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_cancel_ongoing_movements |
| SIPartitionMovementTest  | test_cross_shard                              | {"cloud_storage_type": 1, "num_to_upgrade": 0, "with_cloud_topics": true} | integration | https://buildkite.com/redpanda/redpanda/builds/78409#019b53d4-0da4-4430-832a-f3c6e5305c41 | FLAKY         | 8/11     | Test FAILS after retries.[Significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-B%3A-Test-fails-with-an-improbably-higher-rate-than-baseline-(run-fails))(baseline=0.0058, p0=0.0015, reject_threshold=0.0100)                                                                    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SIPartitionMovementTest&test_method=test_cross_shard                               |
</details>
<!--/body-->

rockwotj commented on src/v/redpanda/admin/services/security.cc:
```diff
@@ -58,7 +58,7 @@ seastar::future<proto::admin::add_role_members_response>
     throw serde::pb::rpc::unimplemented_exception(
-      "AddRoleMembers is not implemented yet"); 
+      "AddRoleMembers is not implemented yet");
 }
```
nit: fixup the clang format changes in dcf4ff2 to the previous commit

rockwotj commented on src/v/redpanda/admin/services/security.cc:
```diff
@@ -12,74 +12,458 @@
+
+    const auto err
+      = co_await _controller->get_security_frontend().local().update_role(
+        role_name, role, model::timeout_clock::now() + role_operation_timeout);
```
probably not a big deal, but there is no consistency guarantees between these operations so it's possible other updates are clobbered in the meantime. The "correct" way to fix this is to have a dedicated controller command for adding/removing members from roles.

Since this isn't possible to do safely today, it's probably OK, but if there is time we should consider following up with a dedicated controller command.

nguyen-andrew replied:
Good point, I've tried to capture this in a Jira ticket here: https://redpandadata.atlassian.net/browse/CORE-15076

rockwotj commented on tests/rptest/tests/rbac_test_v2.py:
```diff
@@ -516,3 +519,213 @@ def test_list_user_roles_no_authn(self):
+    @cluster(num_nodes=3)
+    @parametrize(delete_acls=True)
+    @parametrize(delete_acls=False)
+    @parametrize(delete_acls=None)
```
nit: this turns into 3 different clusters getting spun up and down, consider just having a for loop in the test instead.

rockwotj reviewed (commented):
LGTM, will leave the rest of the review to Mike

nguyen-andrew commented:
[Force push](https://github.com/redpanda-data/redpanda/compare/87927807a824884658ae55848d90d909a61f48c5..0d6d47f57b19e640be2663c430b0c0cd04818251): minor adjustments to address PR comments.

nguyen-andrew commented:
Force pushes:
- [Rebase on latest dev](https://github.com/redpanda-data/redpanda/compare/0d6d47f57b19e640be2663c430b0c0cd04818251..7b6dde579936066448b24584768c57fe9c1eb773)
- [Minor adjustments](https://github.com/redpanda-data/redpanda/compare/7b6dde579936066448b24584768c57fe9c1eb773..802f8d617ba903bdc0d3edebfc997ad4ab9e62ff)
  - Renaming `role_operation_timeout` to `security_operation_timeout`
  - Originally, role names were only validated in CreateRole, but now I've added validation of role names to other role operations too.

nguyen-andrew committed:
admin/tests: unit tests for role conversions

Adds unit tests for the Admin API v2 security service role management
functionality, covering:
- Role name validation (valid names, invalid characters, edge cases)
- Conversion between protobuf role types and security subsystem types
- Round-trip conversion correctness for single/multiple members
- Member deduplication behavior in security roles

These tests validate the transformation layer between the Admin API v2
protobuf interface and the internal security subsystem, ensuring role
and member data is correctly converted in both directions.

nguyen-andrew commented:
[Force push](https://github.com/redpanda-data/redpanda/compare/802f8d617ba903bdc0d3edebfc997ad4ab9e62ff..f65adde9be43afe7c261551a7a6cede60c99d7a7):
- Adding `security_test` unit tests to validate correctness of the protobuf <-> role converters.
- Minor refactoring

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/78409-->
<!--e2d5712d-d441-4241-904b-7fb8d7cafe9f-->
<!--title-->
## Retry command for [Build#78409](https://buildkite.com/redpanda/redpanda/builds/78409)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
skip-redpanda-build
skip-units
skip-rebase
tests/rptest/tests/partition_movement_test.py::SIPartitionMovementTest.test_cross_shard@{"cloud_storage_type":1,"num_to_upgrade":0,"with_cloud_topics":true}
```
<!--/body-->

nguyen-andrew commented:
[Force push](https://github.com/redpanda-data/redpanda/compare/f65adde9be43afe7c261551a7a6cede60c99d7a7..42a0d46d1dc7c57b9ec86956f27ac665dd8432c5) to rebase on latest dev.

michael-redpanda commented on proto/redpanda/core/admin/v2/security.proto:
```diff
@@ -63,10 +134,140 @@ service SecurityService {
 
+// CreateRoleRequest is the request for the CreateRole RPC.
+message CreateRoleRequest {
+    // The Role to create.
+    Role role = 1 [(google.api.field_behavior) = REQUIRED];
```
I can't remember where the discussion landed by can a user provide RoleMembers when creating a role?

nguyen-andrew replied:
Yes, with this approach you can provide RoleMembers when creating the role. The CreateRoleRequest message includes the entire Role resource, so you'd be able to specify its RoleMembers as well. Here's an example:
```
curl -X POST -H "Content-Type: application/json" -H "Authorization: Bearer $TOKEN" --data '{"role": {"name": "role1", "members": [{"user": {"name": "admin"}}, {"user": {"name": "testuser1"}}]}}' 
```

michael-redpanda commented on src/v/redpanda/admin/services/security.cc:
```diff
@@ -12,74 +12,486 @@
+      },
+      [&](const auto&) -> security::role_member {
+          vlog(
+            securitylog.error,
```
nit- warn not error

michael-redpanda commented on src/v/redpanda/admin/services/security.cc:
```diff
@@ -12,74 +12,486 @@
+    const auto& member_type = role_member.type();
+
+    proto::admin::role_member pb_role_member;
+    if (member_type == security::role_member_type::user) {
```
nit: should be a switch statement (I know it' sonly one variant at the moment but there will be more soon)

michael-redpanda commented on tests/rptest/tests/rbac_test_v2.py:
```diff
@@ -0,0 +1,518 @@
+    algorithm = "SCRAM-SHA-256"
+    role_name0 = "foo"
+    role_name1 = "bar"
+    role_name2 = "baz"
```
nit if const should these be UPPER_CASE?

michael-redpanda reviewed (commented):
looks good - I'm not sure we need the last two commits - should you just change the tests to use the v2 endpoint rather than duplicating them?

nguyen-andrew commented:
> looks good - I'm not sure we need the last two commits - should you just change the tests to use the v2 endpoint rather than duplicating them?

@michael-redpanda 
I leaned toward duplicating the ducktape tests so we don’t lose coverage on the existing v1 API while it’s still in use. My thinking was that we can remove the v1 tests once we officially deprecate or remove the v1 API, but until then this keeps coverage intact. Happy to change this if you think replacing them is the better move though.

nguyen-andrew commented:
Force pushes:
- [Validate user role member names to prevent control characters](https://github.com/redpanda-data/redpanda/compare/42a0d46d1dc7c57b9ec86956f27ac665dd8432c5..0b73a016c239f8622d30fa0d2e49158879dc0e02). Added to validate the input to the create role RPC as well as the add/remove role member RPC functions. Added additional tests.
- [Addressed nits](https://github.com/redpanda-data/redpanda/compare/0b73a016c239f8622d30fa0d2e49158879dc0e02..f2413a128b3c6760f7046e472c0174c2f0274286).

nguyen-andrew commented:
[Force push](https://github.com/redpanda-data/redpanda/compare/f2413a128b3c6760f7046e472c0174c2f0274286..0887b37486f8c2c0e1a303c65e801597226adc65) to rebase on latest dev.

nguyen-andrew commented:
[Force push](https://github.com/redpanda-data/redpanda/compare/0887b37486f8c2c0e1a303c65e801597226adc65..c1271c7bf59f7f9f0aa0b18eba3d8d24ef47d0be) again to rebase on actual latest dev (to fix "out-of-date with the base branch" error).

nguyen-andrew commented:
[Force push](https://github.com/redpanda-data/redpanda/compare/c1271c7bf59f7f9f0aa0b18eba3d8d24ef47d0be..2fc835bd88f68cfff84c136989f12711a621eff5) to rebase against dev again... last rebase was on a broken dev.
